components/region_cache_memory_engine/src/range_manager.rs

```diff
@@ -0,0 +1,260 @@
+// Copyright 2024 TiKV Project Authors. Licensed under Apache-2.0.
+
+use std::collections::{BTreeMap, BTreeSet};
+
+use engine_traits::CacheRange;
+
+use crate::engine::{RagneCacheSnapshotMeta, SnapshotList};
+
+#[derive(Debug, Default)]
+pub struct RangeMeta {
+    id: u64,
+    range_snapshot_list: SnapshotList,
+    evicted: bool,
+    can_read: bool,
+    safe_ts: u64,
+}
+
+impl RangeMeta {
+    fn new(id: u64) -> Self {
+        Self {
+            id,
+            range_snapshot_list: SnapshotList::default(),
+            evicted: false,
+            can_read: false,
+            safe_ts: 0,
+        }
+    }
+
+    fn derive_from(id: u64, r: &RangeMeta) -> Self {
+        Self {
+            id,
+            range_snapshot_list: SnapshotList::default(),
+            evicted: false,
+            can_read: r.can_read,
+            safe_ts: r.safe_ts,
+        }
+    }
+
+    pub(crate) fn range_snapshot_list(&self) -> &SnapshotList {
+        &self.range_snapshot_list
+    }
+}
+
+#[derive(Default)]
+struct IdAllocator(u64);
+
+impl IdAllocator {
+    fn allocate_id(&mut self) -> u64 {
+        self.0 += 1;
+        self.0
+    }
+}
+
+// RangeManger manges the ranges for RangeCacheMemoryEngine. Every new ranges
+// (whether created by new_range or by splitted due to eviction) has an unique
+// id so that range + id can exactly locate the position.
+// When an eviction occured, say we now have k1-k10 in self.ranges and the
+// eviction range is k3-k5. k1-k10 will be splitted to three ranges: k1-k3,
+// k3-k5, and k5-k10.
+// k1-k3 and k5-k10 will be new ranges inserted in self.ranges with meta dervied
+// from meta of k1-k10 (only safe_ts and can_read will be derived). k1-k10 will
+// be removed from self.ranges and inserted to self.historical_ranges. Then,
+// k3-k5 will be in the self.evicted_ranges. Now, we cannot remove the data of
+// k3-k5 as there may be some snapshot of k1-k10. After these snapshot are
+// dropped, k3-k5 can be acutally removed.
+#[derive(Default)]
+pub struct RangeManager {
```

Comment:
Add a lock here to guard mutation. SInce I expect this to be read more than written to (e.g., read on any writes), consider using https://rustasync.github.io/runtime/crossbeam/sync/struct.ShardedLock.html

Reply:
It's a field of RangeCacheMemoryEngineCore which is protected by a mutex in RangeCacheMemoryEngine.